internal/resolution/datasource/npm_registry_test.go

```diff
@@ -0,0 +1,141 @@
+package datasource_test
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/google/osv-scanner/internal/resolution/datasource"
+	"github.com/google/osv-scanner/internal/testutility"
+)
+
+type fakeNpmRegistry struct {
+	mu         sync.Mutex
+	repository map[string][]byte // path -> response
+	expectAuth string
+}
+
+func (f *fakeNpmRegistry) setResponse(path string, response []byte) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.repository == nil {
+		f.repository = make(map[string][]byte)
+	}
+	f.repository[path] = response
+}
+
+func (f *fakeNpmRegistry) setExpectAuth(auth string) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.expectAuth = auth
+}
+
+func (f *fakeNpmRegistry) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	path := r.URL.RawPath
+	if path == "" {
+		path = r.URL.Path
+	}
+	f.mu.Lock()
+	wantAuth := f.expectAuth
+	resp, ok := f.repository[strings.TrimPrefix(path, "/")]
+	f.mu.Unlock()
+
+	if wantAuth != "" && r.Header.Get("Authorization") != wantAuth {
+		w.WriteHeader(http.StatusUnauthorized)
+		resp = []byte("unauthorized")
+	} else if !ok {
+		w.WriteHeader(http.StatusNotFound)
+		resp = []byte("not found")
+	}
+
+	if _, err := w.Write(resp); err != nil {
+		log.Fatalf("Write: %v", err)
+	}
+}
+
+func TestNpmRegistryClient(t *testing.T) {
+	t.Parallel()
+
+	//nolint:gosec  // "Potential hardcoded credentials" :)
+	const (
```

Comment:
how are these credentials generated?

Reply:
These are just random base64-encoded strings. `auth` is supposed to be the encoded `username:password`, but I'm really just checking these are included in the Authorization header.

These I just randomly went with `cool:auth` and `nice-token` :slightly_smiling_face: